### PR TITLE
Add declarative pub/sub subscriptions

### PR DESCRIPTION
**Pull request by yaron2**
This PR does the following:

1. Adds the ability to subscribe to topics using a declarative manifest
2. Enables support for Self-Hosted and Kubernetes modes
3. Updates tests

To keep the PR small, there will be a follow-up PR to update the pub/sub e2e test to account for declarative subscriptions.
Issue opened [here](https://github.com/dapr/dapr/issues/2014).

Closes https://github.com/dapr/dapr/issues/1895

### CODE REVIEW

File: pkg/runtime/pubsub/subscriptions_test.go [hunk 1]
```diff
@@ -30,3 +35,42 @@ func TestFilterSubscriptions(t *testing.T) {
...
+		writeSubscriptionToDisk(s, filePath)
+
+		subs := DeclarativeSelfHosted(dir, log)
+		assert.Len(t, subs, 1)
+		assert.Equal(t, "topic1", subs[0].Topic)
+		assert.Equal(t, "myroute", subs[0].Route)
+		assert.Equal(t, "pubsub", subs[0].PubsubName)
+		assert.Equal(t, "scope1", subs[0].Scopes[0])
+	})
+}
```
Comment (mchmarny) [inline pkg/runtime/pubsub/subscriptions_test.go:76]: Add negative/invalid format tests

**Review by mchmarny — COMMENTED** [review]
given the "new" way of configuring subscriptions may be good to add few more test, otherwise lgtm

File: pkg/runtime/pubsub/subscriptions_test.go (hunk 1 above)
Comment (yaron2) [reply pkg/runtime/pubsub/subscriptions_test.go]: Added in latest commit.

**Comment by yaron2** [comment]
> given the "new" way of configuring subscriptions may be good to add few more test, otherwise lgtm

Added.

File: pkg/runtime/runtime.go [hunk 2]
```diff
@@ -794,13 +825,35 @@ func (a *DaprRuntime) getTopicRoutes() (map[string]TopicRoute, error) {
...
+	for _, s := range ds {
+		skip := false
+
+		// don't register duplicate subscriptions
+		for _, sub := range subscriptions {
+			if sub.Route == s.Route && sub.PubsubName == s.PubsubName && sub.Topic == s.Topic {
+				log.Warnf("two identical subscriptions found (sources: declarative, app endpoint). topic: %s, route: %s, pubsubname: %s",
+					s.Topic, s.Route, s.PubsubName)
+				skip = true
+				continue
```
Comment (amanbha) [inline pkg/runtime/runtime.go:848]: break, as we have found a match, no need to continue to compare with rest of the subscirptions.

Comment (yaron2) [reply pkg/runtime/runtime.go]: accepted :) changed in last commit.

**Review by amanbha — APPROVED** [review]

**Review by youngbupark — APPROVED** [review]

**Pull request merged**